POO-Java/LearnToDoRealJava.md

```diff
@@ -55,43 +61,339 @@ public move(int x, int y) {
+    for (size_t i = 0; i < n; ++i) {        //  be modified by this function
+        printf("%d\n", a[i]);
+    }
+}
+
+void zero_array(int* a, size_t n) {   // 'a' isn't marked as const, so it can be
+    for (size_t i = 0; i < n; ++i) {  // modified by this function
+        a[i] = 0;  // mutation occurs here
+    }
+}
+
+int main(void) {
+    int const values[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};  // immutable array
+    print_array(values, 10);  // valid, not modifying 'values'
+    zero_array(values, 10);  // invalid, modifying 'values'
+}
+```
+Curiosamente, esta keyword já existe em C desde 1989. Considerando que Java foi
+criado em 1996, parece que conseguimos regredir.
+
+Adiante, a nossa estratégia aqui é, para cada class que definirmos, definimos
+uma interface que só declara métodos que não modificam a class, e a class
+implementa esta interface.\
+Se a class se chamar `Bar`, vamos chamar à interface `BarView`. É uma *view* de
+Bar. Podemos ver, mas não tocar.\
+Sempre que queremos retornar um `Bar` que não deve ser modificado pelo caller,
+em vez de retornar um `Bar`, retornamos uma `BarView`.
+Aplicando este design ao exemplo anterior, obtemos o seguinte código:
+```java
+interface BarView {  // only declares methods that don't mutate Bar
+    ArrayList<String> get_strings();
+}
+
+class Bar implements BarView {
+    private ArrayList<String> strings;
+
+    public Bar() {
+        this.strings = new ArrayList<>();
+    }
+
+    public ArrayList<String> get_strings() {
+        return (ArrayList<String>)
+            this.strings.clone();  // a clone still happens if this function is
+        }                          // called. We could, in practice, define an
+                                   // ArrayListView, but there's a much simpler
+                                   // solution (to be written soon TM).
+
+    public void clear_strings() {
+        this.strings.clear();
+    }
+}
+
+class Foo {
+    private Bar bar;
+
+    public Foo() {
+        this.bar = new Bar();
+    }
+
+    public BarView get_bar() {  // now returns an immutable view
+        return this.bar;  // implicit cast from Bar to BarView
+    }
+}
+
+// Takes a Bar and computes some result, but does not modify the Bar.
+// Since it doesn't modify the Bar, we can pass a BarView instead.
+void compute(BarView bar) { /* ... */ }
+
+var all_foos = new ArrayList<Foo>(/* ... */);
+for (var foo : all_foos) {
+    compute(foo.get_bar());  // no more unnecessary clones :)
+}
+```
+Conseguimos definir um método de garantir imutabilidade que:
+* não faz cópias desnecessárias, por isso
+* não precisa de lidar com exceções na criação das cópias;
+* não aceita modificações a `Foo.bar`.\
+Por exemplo, tentar chamar `Foo.get_bar().clear_strings()` dá um erro de
+compilação:\
+'`The method clear_strings() is undefined for the type BarView`', que faz todo o
+sentido, porque `BarView` não define esse método.
```

Comment:
you can add here something about kotlin: "This is such a good pattern that kotlin integrated it into it's standard library. [List](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-list/) and [MutableList](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-mutable-list/)"

Reply:
I'm not that familiar with Kotlin, I'd rather have someone who is add that section :eyes:

Otherwise, if you mean literally pasting the quoted text, sure I can do that.

Reply:
Yeah you can just add it as a note like this

```md
*Note:* This is such a good pattern that kotlin, another jvm based language, integrated it into it's standard library. See [List](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-list/) and [MutableList](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-mutable-list/) as some of the many examples of this pattern.
```

Reply:
Done.
By the way, `it's` is a contraction of `it is`.
The possessive pronoun is spelled `its` (of it). 🤓